src/main/java/com/google/api/generator/gapic/composer/grpc/GrpcServiceStubClassComposer.java

```diff
@@ -335,4 +447,120 @@ private LambdaExpr createRequestParamsExtractorClassInstance(Method method) {
         .setReturnExpr(returnExpr)
         .build();
   }
+
+  private void addRequestParamsForRoutingHeders(
+      Method method,
+      List<Statement> classStatements,
+      VariableExpr paramsVarExpr,
+      List<Expr> bodyExprs,
+      VariableExpr requestVarExpr) {
+    ImmutableList<RoutingHeader> routingHeaders = method.routingHeaders().routingHeadersList();
+    for (int i = 0; i < routingHeaders.size(); i++) {
+      RoutingHeader routingHeader = routingHeaders.get(i);
+      MethodInvocationExpr requestFieldGetterExpr =
+          createRequestFieldGetterExpr(requestVarExpr, routingHeader.field());
+      Expr routingHeaderKeyExpr =
+          ValueExpr.withValue(StringObjectValue.withValue(routingHeader.name()));
+      // TODO: Create proper snake style upper case name
+      String pathTemplateName =
+          String.format("PATH_TEMPLATE_%s_%s", method.name().toUpperCase(), i);
+      TypeNode pathTemplateType =
+          TypeNode.withReference(ConcreteReference.withClazz(PathTemplate.class));
+      Variable pathTemplateVar =
+          Variable.builder().setType(pathTemplateType).setName(pathTemplateName).build();
+      Expr routingHeaderPatternExpr = VariableExpr.withVariable(pathTemplateVar);
+      VariableExpr pathTemplateVarExpr =
+          VariableExpr.builder()
+              .setVariable(pathTemplateVar)
+              .setIsDecl(true)
+              .setIsStatic(true)
+              .setIsFinal(true)
+              .setScope(ScopeNode.PRIVATE)
+              .build();
+      ValueExpr valueExpr =
+          ValueExpr.withValue(StringObjectValue.withValue(routingHeader.pattern()));
+      Expr pathTemplateExpr =
+          AssignmentExpr.builder()
+              .setVariableExpr(pathTemplateVarExpr)
+              .setValueExpr(
+                  MethodInvocationExpr.builder()
+                      .setStaticReferenceType(pathTemplateType)
+                      .setMethodName("create")
+                      .setArguments(valueExpr)
+                      .setReturnType(pathTemplateType)
+                      .build())
+              .build();
+      Statement pathTemplateClassVar = ExprStatement.withExpr(pathTemplateExpr);
+      classStatements.add(pathTemplateClassVar);
+      MethodInvocationExpr addParamsMethodExpr =
+          MethodInvocationExpr.builder()
+              .setStaticReferenceType(
+                  TypeNode.withReference(ConcreteReference.withClazz(RoutingHeaderHelper.class)))
+              .setMethodName("addParams")
```

Comment:
Should this be singular? (https://github.com/googleapis/gax-java/pull/1598#discussion_r781574600)

Reply:
Yes, I agree. I'll change it to singular.